### PR TITLE
Fix nodes volumesAttached status not being updated

### LINKED ISSUE #107973: Bug in node_status_updater.go - UpdateNodeStatuses
https://github.com/kubernetes/kubernetes/blob/9b09612d1b0a6ec441833f928cbb205d936da59c/pkg/controller/volume/attachdetach/statusupdater/node_status_updater.go#L92

UpdateNodeStatuses will exit immediately if updateNodeStatus returns an error.
It does mark the node for which the call failed for later update but all the following nodes in the update list are left unchanged and therefore won't be updated.

This can result for instance in a pod getting stuck waiting for its volume to be attached.
Assuming the attach of the volume to the pod's node succeeded but another node preceding this one in the update list has its updateNodeStatus call fail - then the attach detach controller will never update the node status to indicate the volume was attached and the pod will be stuck until there is another update for this node attachements.

The fix can be as simple as using a continue but the original code seems to want to abort the whole loop in the case of that error, so that would require marking all the remaining nodes for update, not just the current one.

### PR DESCRIPTION
**Pull request by jfremy**
/kind bug

Fixes #107973 

The UpdateNodeStatuses code stops too early in case there is an error when calling updateNodeStatus. It will return immediately which means any remaining node won't have its update status put back to true.

Looking at the call sites for UpdateNodeStatuses, it appears this is not the only issue. If the lister call fails with anything but a Not Found error, it's silently ignored which is wrong in the detach path.
Also the reconciler detach path calls UpdateNodeStatuses but the real intent is to only update the node currently processed in the loop and not proceed with the detach call if there is an error updating that specify node volumesAttached property. With the current implementation, it will not proceed if there is an error updating another node (which is not completely bad but not ideal) and worse it will proceed if there is a lister error on that node which means the node volumesAttached property won't have been updated.

To fix those issues, introduce the following changes:
- [node_status_updater] introduce UpdateNodeStatusForNode which does what UpdateNodeStatuses does but only for the provided node
- [node_status_updater] if the node lister call fails for anything but a Not Found error, we will return an error, not ignore it
- [node_status_updater] if the update of a node volumesAttached properties fails we continue processing the other nodes
- [actual_state_of_world] introduce GetVolumesToReportAttachedForNode which does what GetVolumesToReportAttached but for the node whose name is provided it returns a bool which indicates if the node in question needs an update as well as the volumesAttached list. It is used by UpdateNodeStatusForNode
- [reconciler] use UpdateNodeStatusForNode in the detach loop

```release-note
Fix a bug in attachdetach controller that didn't properly handle kube-apiserver errors leading to stuck attachments/detachments.
```

### CODE REVIEW

**Comment by jfremy** [comment]
/retitle Fix nodes volumesAttached status not being updated

File: pkg/controller/volume/attachdetach/cache/actual_state_of_world.go [hunk 1]
```diff
@@ -677,10 +671,49 @@ func (asw *actualStateOfWorld) GetVolumesToReportAttached() map[types.NodeName][
...
+	}
+	if !nodeToUpdateObj.statusUpdateNeeded {
+		return false, nil
+	}
+
+	attachedVolumes = asw.getAttachedVolumeFromUpdateObject(nodeToUpdateObj.volumesToReportAsAttached)
+	// When GetVolumesToReportAttached is called by node status updater, the current status
+	// of this node will be updated, so set the flag statusUpdateNeeded to false indicating
+	// the current status is already updated.
+	if err := asw.updateNodeStatusUpdateNeeded(nodeName, false); err != nil {
```
Comment (gnufied) [inline pkg/controller/volume/attachdetach/cache/actual_state_of_world.go:691]: Looks like we are updating ASW with read lock?

File: pkg/controller/volume/attachdetach/statusupdater/node_status_updater.go [hunk 2]
```diff
@@ -57,40 +60,65 @@ type nodeStatusUpdater struct {
...
-			klog.V(2).Infof("Error retrieving nodes from node lister. Error: %v", err)
-			nsu.actualStateOfWorld.SetNodeStatusUpdateNeeded(nodeName)
-			continue
+		err := nsu.processNodeVolumes(nodeName, attachedVolumes)
+		if err != nil {
+			nodeIssues = append(nodeIssues, nodeName)
 		}
+	}
+	if len(nodeIssues) > 0 {
+		return fmt.Errorf("%v could not be updated", nodeIssues)
```
Comment (gnufied) [inline pkg/controller/volume/attachdetach/statusupdater/node_status_updater.go:74]: We can also return a generic error so as we are not logging 100-300 concatenated node names in logs at once. Something like - `fmt.Errorf("Unable to update %d  nodes for volume attach status")` ? We are anyways logging the error  from individual nodes in `processNodeVolumes` functions.

**Comment by gnufied** [comment]
Thanks for the PR. Some unit tests would be nice which can exercise the bug in a reasonable fashion.

**Comment by jfremy** [comment]
> Thanks for the PR. Some unit tests would be nice which can exercise the bug in a reasonable fashion.

yes will add

File: pkg/controller/volume/attachdetach/cache/actual_state_of_world.go (hunk 1 above)
Comment (jfremy) [reply pkg/controller/volume/attachdetach/cache/actual_state_of_world.go:691]: Good catch - do you want me to fix this as well?

File: pkg/controller/volume/attachdetach/statusupdater/node_status_updater.go (hunk 2 above)
Comment (jfremy) [reply pkg/controller/volume/attachdetach/statusupdater/node_status_updater.go:74]: Sure I can update the messages

**Comment by jfremy** [comment]
> Thanks for the PR. Some unit tests would be nice which can exercise the bug in a reasonable fashion.

Still working on it. There were no unit tests for NodeStatusUpdater so it's a little more work than I expected

**Comment by gnufied** [comment]
One of the ways you can possibly test drive this is by using something like a fakeClient with a custom reactor for stuff. Something like - https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/persistentvolume/testing/testing.go#L554 

There should be other similar examples in the code.

**Comment by jfremy** [comment]
> One of the ways you can possibly test drive this is by using something like a fakeClient with a custom reactor for stuff. Something like - https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/persistentvolume/testing/testing.go#L554 
> 
> 
> 
> There should be other similar examples in the code. 

Yes, that's what I've been using. Initially I was looking at the existing fake client factory used in the other unit tests but it did not really fit the need.
I'll have some unit tests for node status updater soon

**Comment by jfremy** [comment]
> One of the ways you can possibly test drive this is by using something like a fakeClient with a custom reactor for stuff. Something like - https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/persistentvolume/testing/testing.go#L554
> 
> There should be other similar examples in the code.

I've added a few unit tests to check the nominal case and one test that check non-regression on this issue

**Comment by jfremy** [comment]
Here's a summary of the fixes / changes included in this PR:
- fix use of read lock in actual_state_of_world.go GetVolumesToReportAttached - it should be write lock since we modify the map content
- fix bug in node_status_updater.go UpdateNodeStatuses that would return an error immediately if there was an error in node.patch
- fix bug in node_status_updater.go UpdateNodeStatuses that would not return an error and skip the node if the NodeLister returned a non IsNotFound error for the node
- changed NodeStatusUpdater to allow processing a single node - this is used in the detach loop of reconciler.go (which process volumes one by one as opposed to the attach section that does every node/volume in one go) - this required adding an matching method in ActualStateOfWorld

File: pkg/controller/volume/attachdetach/cache/actual_state_of_world.go [hunk 3]
```diff
@@ -677,10 +671,49 @@ func (asw *actualStateOfWorld) GetVolumesToReportAttached() map[types.NodeName][
...
+		klog.Errorf("Failed to update statusUpdateNeeded field when getting volumes: %v", err)
+	}
+
+	return true, attachedVolumes
```
Comment (YuikoTakada) [inline pkg/controller/volume/attachdetach/cache/actual_state_of_world.go:695]: ditto

File: pkg/controller/volume/attachdetach/cache/actual_state_of_world.go [hunk 4]
```diff
@@ -677,10 +671,49 @@ func (asw *actualStateOfWorld) GetVolumesToReportAttached() map[types.NodeName][
 	return volumesToReportAttached
 }
 
+func (asw *actualStateOfWorld) GetVolumesToReportAttachedForNode(nodeName types.NodeName) (bool, []v1.AttachedVolume) {
+	asw.Lock()
+	defer asw.Unlock()
+
+	var attachedVolumes []v1.AttachedVolume
```
Comment (YuikoTakada) [inline pkg/controller/volume/attachdetach/cache/actual_state_of_world.go:678]: nit: `volumesToReportAttached` seems to be easy to understand same as https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/attachdetach/cache/actual_state_of_world.go#L653
```suggestion
	var volumesToReportAttached []v1.AttachedVolume
```

File: pkg/controller/volume/attachdetach/cache/actual_state_of_world.go [hunk 5]
```diff
@@ -677,10 +671,49 @@ func (asw *actualStateOfWorld) GetVolumesToReportAttached() map[types.NodeName][
...
+	nodeToUpdateObj, ok := asw.nodesToUpdateStatusFor[nodeName]
+	if !ok {
+		return false, nil
+	}
+	if !nodeToUpdateObj.statusUpdateNeeded {
+		return false, nil
+	}
+
+	attachedVolumes = asw.getAttachedVolumeFromUpdateObject(nodeToUpdateObj.volumesToReportAsAttached)
```
Comment (YuikoTakada) [inline pkg/controller/volume/attachdetach/cache/actual_state_of_world.go:687]: ditto

**Review by YuikoTakada — COMMENTED** [review]
Thank you for your PR. I prefer this approach. I wrote some comments, so please check them.

**Comment by jfremy** [comment]
Can a trusted user add a retitle command? 
`/retitle Fix nodes volumesAttached status not being updated`

My initial title did not please the bot. I changed it in github but that does not seem to register and I don't have the rights to do the retitle myself

**Comment by gnufied** [comment]
Bot is complaining about commit message in one of your first commits, not the PR title. Can you amend your commit and force push to this branch?

**Comment by jfremy** [comment]
> Bot is complaining about commit message in one of your first commits, not the PR title. Can you amend your commit and force push to this branch?

done. thanks! I find the message from the bot confusing

**Comment by gnufied** [comment]
Thanks for working on it. Can you please squash commits which are not not directly on topic. Generally - commits such as "PR review comments" etc should be squashed before a PR can merge, so as commit history of change looks reasonable.

File: pkg/controller/volume/attachdetach/cache/actual_state_of_world.go [hunk 6]
```diff
@@ -677,10 +671,48 @@ func (asw *actualStateOfWorld) GetVolumesToReportAttached() map[types.NodeName][
 	return volumesToReportAttached
 }
 
+func (asw *actualStateOfWorld) GetVolumesToReportAttachedForNode(nodeName types.NodeName) (bool, []v1.AttachedVolume) {
```
Comment (jingxu97) [inline pkg/controller/volume/attachdetach/cache/actual_state_of_world.go:674]: it seems to me the returned bool is not very necessary. Based on reconciler logic, the only possible case is the node is no longer in actual state. In this case, volume list will be empty. I am wondering in this case, we should update node status to have empty volume list or not. Or Maybe this case of node empty is also should not exist?

Comment (jfremy) [reply pkg/controller/volume/attachdetach/cache/actual_state_of_world.go:674]: If I don't report this bool, then I have no way of knowing if you need to update the node status.volumesAttached property with an empty list or if there is nothing to do (for instance when we detach the last volume from the node)

Comment (jingxu97) [reply pkg/controller/volume/attachdetach/cache/actual_state_of_world.go:674]: What I mean is GetVolumesToReportAttachedForNode() is only needed by UpdateNodeStatusForNode(), which is only called in reconciler during detach, and before calling it, it mark volume/node by RemoveVolumeFromReportAsAttached. In this case, the bool will be always true, right? If this function is used in other places, we need to have bool value, I think.
 https://github.com/kubernetes/kubernetes/pull/108167/files#diff-e76cff1d139f1df6d8c37cea61b19aa8caea6b103d64e58a084f2d81c5bad149R196

**Comment by jingxu97** [comment]
also created an issue to follow up to see whether it is better to simply some of the logic here to avoid future mistakes. https://github.com/kubernetes/kubernetes/issues/108433

@jfremy @gnufied what do you think?

**Comment by mattcary** [comment]
FYI I'll be creating cherrypicks, reach out to me on slack if questions/comments/concerns.

**Comment by gnufied** [comment]
@jingxu97 yeah I think simplifying fetching the node status and update logic makes sense.  Lets do this in a follow up.

**Comment by ialidzhikov** [comment]
> FYI I'll be creating cherrypicks, reach out to me on slack if questions/comments/concerns.

Is this fix applicable for 1.21 as well? I see that currently it is only backported to 1.22 and 1.23.

cc @wojtek-t

**Comment by mattcary** [comment]
I think it would be, but I forget the context. It applies cleanly to 1.21 at least.

\cc @jingxu97 

Do  you remember, Jing?

**Comment by wojtek-t** [comment]
> Is this fix applicable for 1.21 as well? I see that currently it is only backported to 1.22 and 1.23.

Yes - we should backport to 1.21 too.

> I think it would be, but I forget the context. It applies cleanly to 1.21 at least.

Volume attachments/detachments were stuck in case of errors returned from apiserver.
@mattcary - can you please open 1.21 cherrypick since you were opening 1.22 and 1.23 ones?

**Comment by mattcary** [comment]
👍🏾

https://github.com/kubernetes/kubernetes/pull/109206

**Pull request merged**